### PR TITLE
Support unwrap date to timestamp with time zone

### LINKED ISSUE #5798: Optimize date and timestamp(p) with time zone comparisons
See TODO in the code

https://github.com/trinodb/trino/blob/80c079f9254a6786cfe204d617276c6080ae5c2c/core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java#L441-L443

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Fixes #5798

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# General
* Improve performance of specific queries which compare table columns of type 
  `date` with `timestamp with time zone` literals. ({issue}`5798`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 1]
```diff
@@ -436,9 +437,20 @@ private boolean hasInjectiveImplicitCoercion(Type source, Type target, Object va
...
+                    if (!getTimeZone(timestampWithTimeZoneType, value).equals(session.getTimeZoneKey())) {
+                        return false;
+                    }
+
+                    // Cast from DATE to TIMESTAMP WITH TIME ZONE is not monotonic when there is a forward DST change in the session zone
+                    if (!isTimestampToTimestampWithTimeZoneInjectiveAt(session.getTimeZoneKey().getZoneId(), getInstantWithTruncation(timestampWithTimeZoneType, value))) {
+                        return false;
+                    }
+
+                    return true;
```
Comment (ebyhr) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:451]: @findepi I'm not sure about the expected condition here. Should we verify whether all time parts of `value` are zero (e.g. `2022-11-10 00:00:00.000000`)?

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:451]: I don't think it's needes

File: core/trino-main/src/test/java/io/trino/sql/query/TestUnwrapCastInComparison.java [hunk 2]
```diff
@@ -326,6 +326,85 @@ public void testCastTimestampToTimestampWithTimeZone()
...
+
+        for (LocalTime toLocalTime : toLocalTimes) {
+            for (int timestampPrecision : asList(0, 3, 6, 9, 12)) {
+                for (String operator : COMPARISON_OPERATORS) {
+                    validate(
+                            session,
+                            operator,
+                            "date",
+                            "DATE '2017-09-24'",
+                            format("timestamp(%s) with time zone", max(9, timestampPrecision)),
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/sql/query/TestUnwrapCastInComparison.java:401]: why 9? what about testing 12?

File: core/trino-main/src/test/java/io/trino/sql/query/TestUnwrapCastInComparison.java [hunk 3]
```diff
@@ -326,6 +326,85 @@ public void testCastTimestampToTimestampWithTimeZone()
...
+                            format("TIMESTAMP '2017-04-01 %s Z'", toLocalTime));
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/sql/query/TestUnwrapCastInComparison.java:402]: where does "2017-04-01" come from?
should we have tests with some other dates on the RHS?

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 4]
```diff
@@ -436,9 +437,20 @@ private boolean hasInjectiveImplicitCoercion(Type source, Type target, Object va
...
+                }
                 // CAST from TIMESTAMP WITH TIME ZONE to d and back to TIMESTAMP WITH TIME ZONE does not round trip for most types d
-                // TODO add test coverage
-                // TODO (https://github.com/trinodb/trino/issues/5798) handle DATE -> TIMESTAMP WITH TIME ZONE
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:441]: I misunderstood that they're the same items. Left the first TODO comment.

File: core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapCastInComparison.java [hunk 5]
```diff
@@ -728,7 +822,12 @@ private void testUnwrap(String inputType, String inputPredicate, String expected
 
     private void testUnwrap(Session session, String inputType, String inputPredicate, String expectedPredicate)
     {
-        String sql = format("SELECT * FROM (VALUES CAST(NULL AS %s)) t(a) WHERE %s", inputType, inputPredicate);
+        testUnwrap(session, inputType, "NULL", inputPredicate, expectedPredicate);
+    }
+
+    private void testUnwrap(Session session, String inputType, String inputValue, String inputPredicate, String expectedPredicate)
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapCastInComparison.java:828]: Yes, the input value affects it. I left TODO comment as it's a pre-existing issue and the change seems not simple.

File: core/trino-main/src/test/java/io/trino/sql/query/TestUnwrapCastInComparison.java (hunk 3 above)
Comment (ebyhr) [reply core/trino-main/src/test/java/io/trino/sql/query/TestUnwrapCastInComparison.java]: Added other dates to RHS. 
The date comes from the exsiting `testCastTimestampToTimestampWithTimeZone`.

**Comment by ebyhr** [comment]
Addressed comments.

File: core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapCastInComparison.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapCastInComparison.java:828]: i want to understand that. Do you know what happens here?

**Review by findepi — COMMENTED** [review]
LGTM except open https://github.com/trinodb/trino/pull/14945/commits/be819894cf9685265d5287d89e725cb003891072#r1021802073.
(Ideally, there is no  new `testUnwrap` overload.)

File: core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapCastInComparison.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ebyhr) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapCastInComparison.java]: Sorry I was doing something wrong. I reconfirmed the input value doesn't affect it. Removed a new `testUnwrap`

**Comment by findepi** [comment]
Thanks. LGTM

**Pull request merged**